### PR TITLE
feat: add put action

### PR DESCRIPTION
**Pull request by jiuker**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

add put action support -s size -P part_numbers

- [x] avoid the so many call with one upload GetBucketObjectLockConfig/HeadBucket
- [x] avoid mc cp does with complex vairations( source is local path/ target is s3 )

## How to test this PR?

1.deploy a minio
2. mc admin trace -a alias | grep s3.
3. set env MC_REGION=value
4.  `mc put object minio/bucket` , `mc put object minio/bucket/someobject` and `mc put object minio/bucket/dir/`

You will see just upload call here.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fmc%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/client-s3.go [hunk 1]
```diff
@@ -2327,12 +2328,15 @@ func (c *S3Client) objectInfo2ClientContent(bucket string, entry minio.ObjectInf
...
-		return nil, probe.NewError(BucketDoesNotExist{Bucket: bucket})
+	notCheck, _ := strconv.ParseBool(os.Getenv("MC_NOT_CHECK_BUCKET_EXISTS"))
+	if !notCheck {
+		exists, e := c.api.BucketExists(ctx, bucket)
+		if e != nil {
+			return nil, probe.NewError(e)
+		}
+		if !exists {
+			return nil, probe.NewError(BucketDoesNotExist{Bucket: bucket})
+		}
```
Comment (harshavardhana) [inline cmd/client-s3.go:2339]: we can't add random ENVs like this @jiuker. Please revert this change.

File: cmd/put-main.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
```
Comment (harshavardhana) [inline cmd/put-main.go:1]:
```suggestion
// Copyright (c) 2015-2024 MinIO, Inc.
```

File: cmd/put-main.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/dustin/go-humanize"
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/probe"
+)
+
+// put command flags.
+var (
+	putFlags = []cli.Flag{
+		cli.IntFlag{
+			Name:  "P",
```
Comment (harshavardhana) [inline cmd/put-main.go:36]:
```suggestion
			Name:  "parallel, P",
```

File: cmd/put-main.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+			Usage: "parts number",
```
Comment (harshavardhana) [inline cmd/put-main.go:37]:
```suggestion
			Usage: "upload number of parts in parallel",
```

File: cmd/put-main.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		},
+		cli.StringFlag{
+			Name:  "s",
```
Comment (harshavardhana) [inline cmd/put-main.go:40]:
```suggestion
			Name:  "part-size, s",
```

File: cmd/put-main.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+			Usage: "size for each part",
+		},
+	}
+)
+
+// Put command.
+var putCmd = cli.Command{
+	Name:         "put",
+	Usage:        "put local objects to s3",
```
Comment (harshavardhana) [inline cmd/put-main.go:49]:
```suggestion
	Usage:        "upload local objects to object storage",
```

File: cmd/put-main.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	// threads
+	threads := cliCtx.Int("P")
+	if threads == 0 {
+		threads = 1
+	}
+	os.Setenv("MC_UPLOAD_MULTIPART_THREADS", strconv.Itoa(threads))
+	os.Setenv("MC_NOT_CHECK_BUCKET_EXISTS", "TRUE")
+	if os.Getenv("MC_REGION") == "" {
+		os.Setenv("MC_REGION", "us-east-1")
+	}
```
Comment (harshavardhana) [inline cmd/put-main.go:96]:
```suggestion
```

File: cmd/put-main.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+				if !ok {
+					quite <- struct{}{}
+					return
+				}
+				fmt.Println(doCopy(ctx, cpURLs, pg, encKeyDB, false, false, false))
+			}
+		}
+	}()
+	<-quite
+	fmt.Println(errSeen)
```
Comment (harshavardhana) [inline cmd/put-main.go:160]: ?

File: cmd/put-main.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+	//return mainCopy(cliCtx)
```
Comment (harshavardhana) [inline cmd/put-main.go:162]: ?

File: cmd/put-main.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+			}
+
+			totalBytes += cpURLs.SourceContent.Size
+			pg.SetTotal(totalBytes)
+			totalObjects++
+			cpURLsCh <- cpURLs
+		}
+		close(cpURLsCh)
+	}()
+	quite := make(chan struct{})
```
Comment (harshavardhana) [inline cmd/put-main.go:145]:
```suggestion
	quit := make(chan struct{})
```
?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by jiuker** [comment]
```log
MacBook-Pro mc % mc admin trace -a minio | grep s3.
2024-02-04T14:57:57.554 [200 OK] s3.NewMultipartUpload 127.0.0.1:9000/mytest/minio-server?uploads=  127.0.0.1         2.48ms       ↑ 106 B ↓ 315 B
2024-02-04T14:57:57.557 [200 OK] s3.PutObjectPart 127.0.0.1:9000/mytest/minio-server?partNumber=2&uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         268.191ms    ↑ 31 MiB ↓ 0 B
2024-02-04T14:57:57.558 [200 OK] s3.PutObjectPart 127.0.0.1:9000/mytest/minio-server?partNumber=1&uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         270.64ms     ↑ 31 MiB ↓ 0 B
2024-02-04T14:57:57.830 [200 OK] s3.PutObjectPart 127.0.0.1:9000/mytest/minio-server?partNumber=4&uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         317.571ms    ↑ 31 MiB ↓ 0 B
2024-02-04T14:57:57.826 [200 OK] s3.PutObjectPart 127.0.0.1:9000/mytest/minio-server?partNumber=3&uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         324.7ms      ↑ 31 MiB ↓ 0 B
2024-02-04T14:57:58.149 [200 OK] s3.PutObjectPart 127.0.0.1:9000/mytest/minio-server?partNumber=5&uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         31.538ms     ↑ 3.2 MiB ↓ 0 B
2024-02-04T14:57:58.181 [200 OK] s3.CompleteMultipartUpload 127.0.0.1:9000/mytest/minio-server?uploadId=Mzk1MTYyZDAtNDkzYi00MTMyLTk5ZDctYTRlMTYyNzIzZTM3LjI2MDdlMWRhLTRhMmYtNDk3YS1iMDlmLWJkNjM1ZDcxYWY1Yg  127.0.0.1         8.829ms      ↑ 625 B ↓ 315 B

```
@harshavardhana

**Review by vadmeste — COMMENTED** [review]
I am not sure about the exact requirements here, can you explain more @jiuker ? because we already have 'mc cp'. I am not sure why we are adding 'mc put' command here.

**Comment by zveinn** [comment]
Did you run functional-test.sh on this ? 

This PR makes changes to code used by many commands, so we need to be very careful here. I will review this tomorrow.

**Comment by harshavardhana** [comment]
mc cp is costly, mc put is what it is meant for just doing PUTs does not have the burden of cp and it won't make any extra calls other than PUT or multipart upload

File: cmd/put-main.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} [FLAGS] SOURCE [SOURCE...] TARGET
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+ENVIRONMENT VARIABLES:
+  MC_ENCRYPT:      list of comma delimited prefixes
```
Comment (shtripat) [inline cmd/put-main.go:63]: Cant we have flags like `--encrypt` and `--encrypt-key` like `mc cp` command? ENV is fine but additionally we can have command line flags as well.

File: cmd/put-main.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+  MC_ENCRYPT_KEY:  list of comma delimited prefix=secret values
+
+EXAMPLES:
+  01. Put a list of objects from local file system to Amazon S3 cloud storage.
+      {{.Prompt}} {{.HelpName}} Music/*.ogg s3/jukebox/
+`,
+}
+
+// mainPut is the entry point for cp command.
```
Comment (shtripat) [inline cmd/put-main.go:72]:
```suggestion
// mainPut is the entry point for put command.
```

File: cmd/put-main.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	ctx, cancelPut := context.WithCancel(globalContext)
+	defer cancelPut()
+	// part size
+	size := cliCtx.String("s")
+	if size == "" {
+		size = "16mb"
+	}
+	_, perr := humanize.ParseBytes(size)
+	if perr != nil {
+		fatalIf(probe.NewError(perr), "Unable to parse size")
```
Comment (shtripat) [inline cmd/put-main.go:83]:
```suggestion
		fatalIf(probe.NewError(perr), "Unable to parse part size")
```

File: cmd/put-main.go (hunk 11 above)
Comment (zveinn) [reply cmd/put-main.go:63]: We are replacing the encrypt flags with new ones soon, I will add this after that happens.

File: cmd/put-main.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/minio/cli"
+	"github.com/minio/mc/pkg/probe"
+)
+
+// put command flags.
+var (
+	putFlags = []cli.Flag{
```
Comment (shtripat) [inline cmd/put-main.go:33]: any reason to keep one short flag uppercase and one lowercase :)

File: cmd/put-main.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+			sourceURLs:         sourceURLs,
+			targetURL:          targetURL,
+			isRecursive:        false,
+			encKeyDB:           encKeyDB,
+			olderThan:          "",
+			newerThan:          "",
+			timeRef:            time.Time{},
+			versionID:          "",
+			isZip:              false,
+			ignoreBucketExists: true,
```
Comment (shtripat) [inline cmd/put-main.go:122]: somehow I feel `checkBucketExists` as better name or else `ignoreBucketExistsCheck`

**Review by shtripat — COMMENTED** [review]
Few minor ones. command works as expected though.

File: cmd/put-main.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+			Name:  "part-size, s",
+			Usage: "each part size",
+		},
+	}
+)
+
+// Put command.
+var putCmd = cli.Command{
+	Name:         "put",
+	Usage:        "upload local objects to object storage",
```
Comment (shtripat) [inline cmd/put-main.go:47]:
```suggestion
	Usage:        "upload local object to s3 object storage",
```
Removed plural for `object` as at a time we can upload/put one object

File: cmd/put-main.go [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+ENVIRONMENT VARIABLES:
+  MC_ENCRYPT:      list of comma delimited prefixes
+  MC_ENCRYPT_KEY:  list of comma delimited prefix=secret values
+
+EXAMPLES:
+  01. Put a list of objects from local file system to Amazon S3 cloud storage.
```
Comment (shtripat) [inline cmd/put-main.go:66]: I dont think we can put a list of objects in one shot so we should rephrase the text here. Better help examples could be
```
1. Put an object from local file system to S3 storage
{{.Prompt}} {{.HelpName}} path-to/object ALIAS/BUCKET
2. Put an object from local file system to S3 bucket with name
{{.Prompt}} {{.HelpName}} path-to/object ALIAS/BUCKET/OBJECT-NAME
3. Put an object from local file system to S3 bucket under a prefix
{{.Prompt}} {{.HelpName}} path-to/object ALIAS/BUCKET/PREFIX/
```
just suggestion if looks good.

**Review by shtripat — APPROVED** [review]
lgtm

File: cmd/put-main.go [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/minio/mc/pkg/probe"
+)
+
+// put command flags.
+var (
+	putFlags = []cli.Flag{
+		cli.IntFlag{
+			Name:  "parallel, P",
+			Usage: "upload number of parts in parallel",
+		},
```
Comment (harshavardhana) [inline cmd/put-main.go:36]:
```suggestion
		cli.IntFlag{
			Name:  "parallel, P",
			Usage: "upload number of parts in parallel",
			Value: 4,
		},
```

Must have a default value set to '4'

File: cmd/put-main.go [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+		cli.StringFlag{
+			Name:  "part-size, s",
+			Usage: "each part size",
+		},
```
Comment (harshavardhana) [inline cmd/put-main.go:40]:
```suggestion
		cli.StringFlag{
			Name:  "part-size, s",
			Usage: "each part size",
			Value: "16MiB",
		},
```

Must have a default value.

File: cmd/put-main.go [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+		size = "16mb"
+	}
+	_, perr := humanize.ParseBytes(size)
+	if perr != nil {
+		fatalIf(probe.NewError(perr), "Unable to parse part size")
+	}
+	os.Setenv("MC_UPLOAD_MULTIPART_SIZE", size)
+	// threads
+	threads := cliCtx.Int("P")
+	if threads == 0 {
```
Comment (harshavardhana) [inline cmd/put-main.go:91]: validate for threads < 0

File: cmd/put-main.go [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+	// part size
+	size := cliCtx.String("s")
+	if size == "" {
+		size = "16mb"
+	}
+	_, perr := humanize.ParseBytes(size)
+	if perr != nil {
+		fatalIf(probe.NewError(perr), "Unable to parse part size")
+	}
+	os.Setenv("MC_UPLOAD_MULTIPART_SIZE", size)
```
Comment (harshavardhana) [inline cmd/put-main.go:88]:
```suggestion
	os.Setenv("MC_UPLOAD_MULTIPART_SIZE", size)
```

This is not an expected pattern of passing values; please feel free to make relevant changes to the function doCopy() to pass the information pertinent to the s3Client.

You do not have to re-use the ENVs in this manner; this is not the correct style that we follow in the MC codebase.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**